src/main/kotlin/nl/info/zac/authentication/UserPrincipalFilter.kt

```diff
@@ -76,9 +89,67 @@ constructor(
                     setLoggedInUserOnHttpSession(userPrincipal as OidcPrincipal<*>, httpSession)
                 }
         }
+
+        if (!isAuthorizationAllowed(request)) {
+            response.sendError(jakarta.servlet.http.HttpServletResponse.SC_FORBIDDEN)
+            return
+        }
+
         filterChain.doFilter(servletRequest, servletResponse)
     }
 
+    @Suppress("ReturnCount")
+    private fun isAuthorizationAllowed(request: HttpServletRequest): Boolean {
+        val session = request.getSession(false) ?: return true
+        val user = getLoggedInUser(session) ?: return true
+
+        val path = request.requestURI.removePrefix(request.contextPath ?: "")
+        val isAdmin = adminUriPrefixes.any { prefix -> path.startsWith(prefix) }
+
+        return if (pabcIntegrationEnabled) {
+            if (isAdmin) {
+                // PABC enabled: admin allowed if the user is authorized for ALL entity types
+                isAuthorizedForAllEntityTypes(user)
+            } else {
+                // PABC enabled: access allowed if the user has >=1 app role on >=1 zaaktype
+                hasAnyPabcApplicationRole(user)
+            }
+        } else {
+            // PABC disabled: legacy web.xml rules on token roles
+            if (isAdmin) {
+                user.roles.contains("beheerder")
+            } else {
+                user.roles.any { it in setOf("raadpleger", "behandelaar", "coordinator", "recordmanager", "beheerder") }
```

Comment:
Please use constants for the ZAC application roles. If we don't have them already I would add them to the `ZACRole` enum. There we should make `DOMEIN_ELK_ZAAKTYPE` deprecated because it is only used in the old IAM architecture.